implement lookup of salt master via dns SRV record

Crapworks opened the pull request:
This PR implements the discovery of  one or multiple salt master via dns SRV records.

This has been asked for here #20275.

I was trying to implement this with the absolute minimum set of changes to not mess anything up. If I was missing obvious things that would cause problems, please let me know so I can implement them.

Commit:
implement lookup of salt master via dns SRV record

cachedout commented:
Wow, you're quick! Thanks very much for this, @Crapworks.

Once the tests finish running, I will take a careful look at this and hopefully get it in. Thanks so much for taking this on. This is a great addition!

Crapworks commented:
One of linting checks was complaining. I've fixed the styleguide errors, but the problem ist that `subprocess.check_output()` is only available for python2.7. Is there a helper function somewhere that I can use in `minion.py` that retuns stdout of an external command?

cachedout commented:
Use `self.functions['cmd.stdout']()` which will made to the `stdout` function in salt.modules.cmdmod`. You might have to move that function around in a bit so it's a method in one of the minion classes that has `functions` populated.

Crapworks commented:
A quick look told me that apparently `eval_master()` is called before `gen_modules()` is populating `self.functions`. I'll have a closer look and see if I can use it or if I have to go the (not so funny) `communicate()` way of the `subprocess` module.

cachedout commented:
I confess to not looking at the code before commenting. :] Apologies. We do have a few things in `salt.utils` like `nb_open` or `timed_subprocess` you could use as well.

cachedout commented:
@Crapworks Did you want to get back to this one? Please let me know if you've had a chance to explore those other alternatives mentioned.

cachedout commented:
Go Go Jenkins!

cachedout commented:
Nitrogen will target Python 2.7 and above so the issue we had with this earlier should no longer be present. We ought to be able to get this in.

tbaker57 commented:
Does that mean Nitrogen won't be supported on Centos/RHEL 6 when it comes out ?

Cheers,
Tim

cachedout commented:
@tbaker57 I knew that question was coming. ;] We'll get out a post out to salt-users soon, but the plan right now is to package and distribute a version of Python that will be supported.

tbaker57 commented:
Whew! :smile:

bkruger99 commented:
You'll want to be careful with this for cross-os compatibility.  For instance, FreeBSD 10 and 11 no longer include 'dig', but have a compatible command called 'drill'.

bkruger99 commented on salt/minion.py:
```diff
@@ -198,6 +199,20 @@ def resolve_dns(opts, fallback=True):
+def srv_lookup(name, protocol='TCP', domain=None):
+    if domain:
+        domain = '.' + domain
+    record_name = '_{0}_._{1}{2}'.format(name, protocol.lower(), domain or '')
```
Format is: _service._proto.name.

Just need to remove the single '_' after {0}.

bkruger99 commented on salt/minion.py:
```diff
@@ -198,6 +199,20 @@ def resolve_dns(opts, fallback=True):
+    if domain:
+        domain = '.' + domain
+    record_name = '_{0}_._{1}{2}'.format(name, protocol.lower(), domain or '')
+    cmd = ['dig', '+short', record_name, 'SRV']
```
host equiv would be:

cmd = ['host', '-t', 'SRV', record_name]

The regex seems to work fine when using this.

bkruger99 commented on salt/minion.py:
```diff
@@ -198,6 +199,20 @@ def resolve_dns(opts, fallback=True):
+        raise err
+    else:
+        regex = re.compile('(?P<prio>\d+) (?P<weight>\d+) (?P<port>\d+) (?P<host>.+)')
+        return [match.groupdict() for match in regex.finditer(result)]
```
If we care about ordering in the record itself, then something along the lines of:

```
records = [match.groupdict() for match in regex.finditer(result)]
return sorted(records, key=lambda k: (int(k['prio']), -int(k['weight'])))

```

This would return records based on priority (lowest preferred) and then weight (higher preferred) .

Commit:
Updates for the salt-stack master poller based off SRV record.
Follow spec, try multiple ways to do the lookup

Crapworks committed:
Merge pull request #1 from bkruger99/dnsdico

Updates for the salt-stack master poller based off SRV record.

bkruger99 commented:
Don't merge yet if it passes, one more thing should be added, i'll do that today..  Should not lock in the name 'saltmaster' in case of a scenario where there are multiple salt masters in different datacenters, but under the same domain name.

Commit:
Add new option of: master_dns_discovery_name, in case we want to use multiple independent masters with a single domain
Add config to minion default config file.
Fix priority / weighting, incorrect statement got placed in first commit
Add debug logging a couple of places
add defaults into the __init__ for the minion options
add some error checking and exit if we have problems

Crapworks committed:
Merge pull request #2 from bkruger99/dnsdico

Add new option of: master_dns_discovery_name, in case we want to use …

cachedout commented:
@bkruger99 Sounds good. Keep me posted and thanks!. :]

bkruger99 commented:
@cachedout I submitted my pull request @Crapworks 's repo and he merged it in.  I see it in this pull request, review at your leisure and let me know what changes need to be made if any..

cachedout commented:
Go Go Jenkins!

rallytime commented:
@Crapworks Apologies for not getting back to this PR sooner. We'd love to keep this moving. The first thing I'd like to get this PR going again is to get a more recent test run on this PR. Would you be willing to rebase this on the latest version of develop? Then we can get an accurate run and make sure this still all works with the current code base.

rallytime commented:
Since we didn't hear back on this change, I am going to close this PR. We can of course re-open this when you're ready to return to it. It will need a rebase against the current version of `develop` and the merge conflicts will need to be addressed as well. Or you can or course open a new PR. Thank you!
